src/topology/test/backpressure.rs

```diff
@@ -0,0 +1,273 @@
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+
+use tokio::time::Duration;
+
+use crate::config::SinkOuter;
+
+use crate::topology::builder::PIPELINE_BUFFER_SIZE;
+use crate::{config::Config, test_util::start_topology};
+
+use vector_core::buffers::{BufferConfig, BufferType, WhenFull};
+
+pub const MEMORY_BUFFER_DEFAULT_MAX_EVENTS: usize = 500;
+
+#[tokio::test]
+async fn serial_backpressure() {
+    let mut config = Config::builder();
+
+    let events_to_sink = 100;
+
+    let expected_sourced_events =
+        events_to_sink + MEMORY_BUFFER_DEFAULT_MAX_EVENTS + PIPELINE_BUFFER_SIZE + 3;
```

Comment:
Where does the `3` come from?

Reply:
I think this comes from the fact that each `Sender` gets 1 additional buffer, and the way we hook stuff up does this 3 times.
https://docs.rs/futures/0.3.19/futures/channel/mpsc/fn.channel.html

That does mean that this test might break if we change how things are hooked up, but I don't think that's the worst thing. It forces you to review exactly what you changed and make sure it was intentional.

Reply:
Gotcha, that makes sense. It might be worth leaving a comment about that.